src/services/pallets/PalletsEventsService.ts

```diff
@@ -0,0 +1,122 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import { ApiDecoration } from '@polkadot/api/types';
+import { Text } from '@polkadot/types';
+import { BlockHash, EventMetadataLatest } from '@polkadot/types/interfaces';
+import { IsEvent } from '@polkadot/types/metadata/decorate/types';
+import { AnyTuple } from '@polkadot/types-codec/types';
+import { stringCamelCase } from '@polkadot/util';
+import { IPalletEvents, IPalletEventsItem } from 'src/types/responses';
+
+import { AbstractPalletsService } from '../AbstractPalletsService';
+
+interface IFetchPalletArgs {
+	hash: BlockHash;
+	palletId: string;
+}
+
+interface IFetchEventItemArgs extends IFetchPalletArgs {
+	eventItemId: string;
+	metadata: boolean;
+}
+
+export class PalletsEventsService extends AbstractPalletsService {
+	async fetchEventItem(
+		historicApi: ApiDecoration<'promise'>,
+		{ hash, palletId, eventItemId, metadata }: IFetchEventItemArgs
+	): Promise<IPalletEventsItem> {
+		const metadataFieldType = 'events';
+		const palletMetadata = historicApi.registry.metadata;
+
+		const [palletMeta, palletMetaIdx] = this.findPalletMeta(
+			palletMetadata,
+			historicApi,
+			palletId,
+			metadataFieldType
+		);
+
+		// Even if `eventItemMeta` is not used, we call this function to ensure it exists. The side effects
+		// of the event item not existing are that `findPalletFieldItemMeta` will throw.
+		const eventItemMetadata = this.findPalletFieldItemMeta(
+			historicApi,
+			palletMeta,
+			eventItemId,
+			metadataFieldType
+		) as EventMetadataLatest;
+
+		let palletEventMetadata: EventMetadataLatest | undefined;
+		if (metadata) {
+			palletEventMetadata = (eventItemMetadata[1] as IsEvent<AnyTuple>).meta;
+		}
+
+		const { number } = await this.api.rpc.chain.getHeader(hash);
+
+		return {
+			at: {
+				hash: hash,
+				height: number.unwrap().toString(10),
+			},
+			pallet: stringCamelCase(palletMeta.name),
+			palletIndex: palletMetaIdx,
+			eventItem: eventItemId,
+			metadata: palletEventMetadata,
+		};
+	}
+
+	async fetchEvents(
+		historicApi: ApiDecoration<'promise'>,
+		{ hash, palletId, onlyIds }: IFetchPalletArgs & { onlyIds: boolean }
+	): Promise<IPalletEvents> {
+		const metadataFieldType = 'events';
+		const metadata = historicApi.registry.metadata;
+		const [palletMeta, palletMetaIdx] = this.findPalletMeta(
+			metadata,
+			historicApi,
+			palletId,
+			metadataFieldType
+		);
+
+		const { number } = await this.api.rpc.chain.getHeader(hash);
+		const parsedPalletName = stringCamelCase(palletMeta.name.toString());
+		const events = historicApi.events[parsedPalletName];
+
+		let items: [] | EventMetadataLatest[] | Text[];
+		if ((palletMeta.events as unknown as EventMetadataLatest).isEmpty) {
+			items = [];
+		} else if (onlyIds) {
+			items = Object.entries(events).map(
+				(eventItem) => eventItem[0] as unknown as Text
+			);
+		} else {
+			items = [] as EventMetadataLatest[];
```

Comment:
Is this line a repeat, can we just reference `items` on line 97?

Reply:
originally added 105 because the compiler complained that items was being used before being initialized without it. Ill refactor this.